src/modules/prpn/blocks.py

```diff
@@ -0,0 +1,48 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+def stick_breaking(logits):
+    e = F.sigmoid(logits)
+    z = (1 - e).cumprod(dim=1)
+    p = torch.cat([e.narrow(1, 0, 1), e[:, 1:] * z[:, :-1]], dim=1)
+
+    return p
+
+
+def softmax(x, mask=None):
```

Comment:
This is a bit odd. There is _definitely_ an efficient implementation of this built into PyTorch...

Reply:
They modify the built-in softmax to include masking for self-attention.